lib/puppet/agent.rb

```diff
@@ -60,8 +63,17 @@ def run(client_options = {})
               end
             end
           rescue Puppet::LockError
-            Puppet.notice _("Run of %{client_class} already in progress; skipping  (%{lockfile_path} exists)") % { client_class: client_class, lockfile_path: lockfile_path }
-            nil
+            wait_for_lock_deadline ||= Time.now.to_i + maxwaitforlock
+
+            if Time.now.to_i > wait_for_lock_deadline
+              Puppet.notice _("Exiting now because the maximum wait for lock time has been exceeded.")
+              nil
+            else
+              Puppet.info _("Another puppet instance is already running")
+              Puppet.info _("Waiting for it to finish. Will try again in %{time} seconds.") % {time: waitforlock}
+              sleep waitforlock
+              retry
+            end
```

Comment:
This is a light version of the `wait for lock` mechanism (see [PUP-8051](https://tickets.puppetlabs.com/browse/PUP-8051) with hardcoded values. If that will be properly done for 5.5.x, this code should be adapted accordingly then. Otherwise, I think the values should be carefully chosen.